lib/Dialect/Debug/DebugOps.cpp

```diff
@@ -0,0 +1,128 @@
+//===- DebugOps.cpp - Debug dialect operations ----------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Dialect/Debug/DebugOps.h"
+#include "mlir/IR/OpImplementation.h"
+
+using namespace circt;
+using namespace debug;
+using namespace mlir;
+
+//===----------------------------------------------------------------------===//
+// StructOp
+//===----------------------------------------------------------------------===//
+
+ParseResult StructOp::parse(OpAsmParser &parser, OperationState &result) {
+  // Parse the struct fields.
+  SmallVector<Attribute> names;
+  SmallVector<OpAsmParser::UnresolvedOperand, 16> operands;
+  std::string nameBuffer;
+  auto parseField = [&]() {
+    nameBuffer.clear();
+    if (parser.parseString(&nameBuffer) || parser.parseColon() ||
+        parser.parseOperand(operands.emplace_back()))
+      return failure();
+    names.push_back(StringAttr::get(parser.getContext(), nameBuffer));
+    return success();
+  };
+  if (parser.parseCommaSeparatedList(AsmParser::Delimiter::Braces, parseField))
+    return failure();
+  assert(names.size() == operands.size());
```

Comment:
This should probably either be an actual error or let verification catch it. Sometimes the latter produces bad errors, e.g., stack traces.

Reply:
The lambda above always pushes into both lists (`operands.emplace_back()` and `names.push_back()`). Maybe I could just drop the assert, since the thing it checks is just in the lines above and hard to break :grin: